core/src/message/user.rs

```diff
@@ -0,0 +1,276 @@
+
+// Copyright (C) 2022-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    ids::{MessageId, ProgramId},
+    message::{Payload, ReplyDetails, Value},
+};
+use alloc::string::ToString;
+use core::convert::TryFrom;
+use gear_core_errors::ReplyCode;
+use scale_info::{
+    scale::{Decode, Encode},
+    TypeInfo,
+};
+
+use super::{MessageDetails, StoredMessage};
+
+/// Message sent to user and deposited as event.
+#[derive(Clone, Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct UserMessage {
+    /// Message id.
+    id: MessageId,
+    /// Message source.
+    source: ProgramId,
+    /// Message destination.
+    destination: ProgramId,
+    /// Message payload.
+    payload: Payload,
+    /// Message value.
+    #[codec(compact)]
+    value: Value,
+    /// Message details: reply message ID and reply code if exists.x
+    details: Option<ReplyDetails>,
+}
+
+impl UserMessage {
+    /// Create new UserMessage.
+    pub fn new(
+        id: MessageId,
+        source: ProgramId,
+        destination: ProgramId,
+        payload: Payload,
+        value: Value,
+        details: Option<ReplyDetails>,
+    ) -> Self {
+        Self {
+            id,
+            source,
+            destination,
+            payload,
+            value,
+            details,
+        }
+    }
+
+    /// Message id.
+    pub fn id(&self) -> MessageId {
+        self.id
+    }
+
+    /// Message source.
+    pub fn source(&self) -> ProgramId {
+        self.source
+    }
+
+    /// Message destination.
+    pub fn destination(&self) -> ProgramId {
+        self.destination
+    }
+
+    /// Message payload bytes.
+    pub fn payload_bytes(&self) -> &[u8] {
+        self.payload.inner()
+    }
+
+    /// Message value.
+    pub fn value(&self) -> Value {
+        self.value
+    }
+
+    /// Message reply details.
+    pub fn details(&self) -> Option<ReplyDetails> {
+        self.details
+    }
+
+    #[allow(clippy::result_large_err)]
+    /// Consumes self in order to create new `StoredMessage`, which payload
+    /// contains string representation of initial bytes,
+    /// decoded into given type.
+    pub fn with_string_payload<D: Decode + ToString>(self) -> Result<Self, Self> {
```

Comment:
Maybe good for now, but create issue to remove converting to string representation, for runtime at least